File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.*;
import java.util.StringTokenizer;


public class Main {
    static int[] arr;
    static int N, S, cnt;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());

        arr = new int[N];
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        dfs(0, 0);
        if(S==0) cnt--;
        System.out.println(cnt);
    }

    static void dfs(int depth, int sum){
        if(depth==N) {
            if(sum==S)
                cnt++;
            return;
        }
        dfs(depth+1, sum+arr[depth]);
        dfs(depth+1, sum);
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import sys

N, S = map(int, sys.stdin.readline().split())
numbers = list(map(int, sys.stdin.readline().split()))

cnt = 0
def recur(sum_, index):
    global N, S, cnt

    if index==N: # 모든 원소를 다 탐색했을 때
        if sum_ == S: # 합이 S인 부분 수열을 찾음
            cnt += 1
        return

    # 다음 인접한 숫자 선택하러 가기
    recur(sum_+numbers[index], index + 1) # 현재 원소를 선택하는 경우
    recur(sum_, index+1) # 현재 원소를 선택하지 않는 경우

recur(0, 0)

if S==0: # S가 0일 때 공집합이 포함되므로 카운트에서 제외
    cnt-=1
print(cnt)